go-controller/pkg/node/gateway_localnet.go

```diff
@@ -239,88 +246,357 @@ func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
+				klog.Errorf("error adding routing table entry for ExternalIP %s: stdout: %s, stderr: %s, err: %v", externalIP, stdout, stderr, err)
+			}
+		}
 	}
-	rules := localnetIptRules(svc, npw.gatewayIP)
-	klog.V(5).Infof("Add rules %v for service %v", rules, svc.Name)
-	return addIptRules(npw.ipt, rules)
+	klog.V(5).Infof("add iptables rules: %v for service: %v", iptRules, svc.Name)
+	return addIptRules(npw.ipt, iptRules)
 }
 
+// Cases to consider here:
+// 1) The effects of 4) from addService means that we only need to look at activeSockets when deleting iptables rules.
 func (npw *localnetNodePortWatcherData) deleteService(svc *kapi.Service) error {
-	if !util.ServiceTypeHasNodePort(svc) {
-		return nil
+	iptRules := []iptRule{}
+	for _, port := range svc.Spec.Ports {
+		if util.ServiceTypeHasNodePort(svc) {
+			if socket, exists := activeSockets[port.NodePort]; exists {
+				if err := socket.Close(); err != nil {
+					klog.Errorf("error closing socket for NodePort svc: %s on port: %v, err: %v", svc.Name, port, err)
+				}
+				delete(activeSockets, port.NodePort)
+				iptRules = append(iptRules, npw.nodePortIPTRules(port, npw.gatewayIP)...)
+			}
+		}
+		for _, externalIP := range svc.Spec.ExternalIPs {
+			if _, exists := npw.localAddrSet[externalIP]; exists {
+				if socket, exists := activeSockets[port.Port]; exists {
+					if err := socket.Close(); err != nil {
+						klog.Errorf("error closing socket for ExternalIP svc: %s on port: %v, err: %v", svc.Name, port, err)
+					}
+					delete(activeSockets, port.Port)
+					iptRules = append(iptRules, npw.externalIPTRules(port, externalIP, svc.Spec.ClusterIP)...)
+					continue
+				}
+			}
+			if npw.isNetworkConnected(net.ParseIP(externalIP)) {
+				continue
+			}
+			if stdout, stderr, err := util.RunIP("route", "del", externalIP, "via", npw.gatewayIP, "dev", localnetGatewayNextHopPort, "table", localnetGatwayExternalIDTable); err != nil {
+				klog.Errorf("error delete routing table entry for ExternalIP %s: stdout: %s, stderr: %s, err: %v", externalIP, stdout, stderr, err)
+			}
+		}
 	}
-	rules := localnetIptRules(svc, npw.gatewayIP)
-	klog.V(5).Infof("Delete rules %v for service %v", rules, svc.Name)
-	delIptRules(npw.ipt, rules)
-	return nil
+	klog.V(5).Infof("delete iptables rules: %v for service: %v", iptRules, svc.Name)
+	return delIptRules(npw.ipt, iptRules)
 }
 
-func localnetNodePortWatcher(ipt util.IPTablesHelper, wf *factory.WatchFactory, gatewayIP net.IP) error {
-	// delete all the existing OVN-NODEPORT rules
-	// TODO: Add a localnetSyncService method to remove the stale entries only
-	_ = ipt.ClearChain("nat", iptableNodePortChain)
-	_ = ipt.ClearChain("filter", iptableNodePortChain)
+func (npw *localnetNodePortWatcherData) syncServices(serviceInterface []interface{}) {
+	removeStaleIPTRules := func(table, chain string, serviceRules []iptRule) {
```

Comment:
its not obvious (at least to me) what serviceRules means here. After reading the code it means "rules that should exist", but its not obvious otherwise. Maybe consider a variable name that describes "these should exist" and not "these should be removed"

Reply:
Agree, those names are not clear enough. I will `s/serviceRules/keepIPTRules/g`